media/server/main/source/MainThread.cpp

```diff
@@ -160,12 +160,31 @@ void MainThread::enqueueTaskAndWait(uint32_t clientId, Task task)
         std::unique_lock<std::mutex> lockTask(*(newTask->mutex));
         {
             std::unique_lock<std::mutex> lockQueue(m_taskQueueMutex);
-            m_taskQueue.push(newTask);
+            m_taskQueue.push_back(newTask);
         }
         m_taskQueueCv.notify_one();
 
         newTask->cv->wait(lockTask);
     }
 }
 
+void MainThread::enqueuePriorityTaskAndWait(uint32_t clientId, Task task)
+{
+    std::shared_ptr<TaskInfo> newTask = std::make_shared<TaskInfo>();
+    newTask->clientId = clientId;
+    newTask->task = task;
+    newTask->mutex = std::make_unique<std::mutex>();
+    newTask->cv = std::make_unique<std::condition_variable>();
+
+    {
+        std::unique_lock<std::mutex> lockTask(*(newTask->mutex));
+        {
+            std::unique_lock<std::mutex> lockQueue(m_taskQueueMutex);
+            m_taskQueue.push_front(newTask);
```

Comment:
Shouldn't we use use priority queue instead? Right now it's only one task with higher priority, so it's not a problem, but if we would like to prioritise more tasks in the future, then prio queue would be more suitable in my opinion.

Reply:
Since we only have 2 levels of priority then I think that priority queue isn't necessary (for ANY number of tasks). A double ended queue is simple and as fast as it could be?

(Priority queue just optimises fetching the largest or smallest from a collection, so we'd have to make every task aware of it's own priority for sorting purposes)

Reply:
The problem isn't that tasks with different priorities would be mixed up, but that high priority tasks would be scheduled in LIFO and low prio tasks would be scheduled in FIFO. Using LIFO for high priority tasks could lead to nasty race conditions.
I already talked with Marcin about that and we are guarded by blocked threads, but anyway I would change to prio queue for readability.

Reply:
as discussed on slack, it's not possible